### PR TITLE
Clean the env vars of all TF_* flags

### PR DESCRIPTION
**Pull request by eparis**
Such variables would allow a crafty user to directly manipulate terraform.
We do not want users doing that. So clear those env vars.

### CODE REVIEW

**Comment by stbenjam** [comment]
Would it be possible to whitelist TF_LOG?

File: cmd/openshift-install/main.go [hunk 1]
```diff
@@ -40,6 +40,19 @@ func main() {
...
 }
 
+// Drop all terraform environment variables that the user may have passed
+// so they can't use terraform directly.
+func cleanEnvironment() {
+	environ := os.Environ()
+	for _, env := range environ {
+		splits := strings.Split(env, "=")
+		key := splits[0]
+		if strings.HasPrefix(key, "TF_") {
```
Comment (zeenix) [inline cmd/openshift-install/main.go]: Doesn't this also unset `TF_VAR_libvirt_master_memory` and `TF_VAR_libvirt_master_vcpu`? We need a way to do that.

File: cmd/openshift-install/main.go [hunk 2]
```diff
@@ -40,6 +40,19 @@ func main() {
...
+			os.Unsetenv(key)
```
Comment (cgwalters) [inline cmd/openshift-install/main.go]: Is this called early enough?  Environment mutation is [unsafe in the presence of threads](https://github.com/rust-lang/rust/pull/24741).

File: cmd/openshift-install/main.go (hunk 1 above)
Comment (eparis) [reply cmd/openshift-install/main.go]: If the memory requirements are inadequate you need to fix the template. Terraform is not for users.

**Comment by abhinavdahiya** [comment]
The installer use aws-sdk, which allow certains envs that can be used to turn on debugging etc.. So i'm not sure why this is being forced?

This only allows vars to changes at global level, which is already very selective... so i'm not sure how _preventive_ this is.

and we we do want to do that, can we move this to `pkg/terraform/exec` before we execute the terraform bindings.

TF_LOG needs to be allowed.

File: cmd/openshift-install/main.go (hunk 1 above)
Comment (zeenix) [reply cmd/openshift-install/main.go]: Yeah, we can have the right amount for memory in the template i guess. What about the VCPU? I guess it's time we implement suggestion of @cgwalters to autodetect that.

**Comment by abhinavdahiya** [comment]
Looking at https://www.terraform.io/docs/configuration/variables.html#assigning-values-to-root-module-variables
```
When you declare variables in the root module of your configuration, you can set their values using CLI options and environment variables. When you declare them in child modules, the calling module should pass values in the module block.
```

the env can only be used to set the variable of the **root** module like [variables-aws.tf](https://github.com/openshift/installer/blob/e1e3c787398632a227df9b0598095536cf2e3547/data/data/aws/variables-aws.tf), which are already exposed through install-config to the user.

So i don't see the use-case for removing the envs.

/close

**Comment by Gal-Zaidman** [comment]
We are using TF_VAR_* on ovirt-e2e job, so this will always fail on our CI.
Is this really needed?

**Comment by abhinavdahiya** [comment]
> https://github.com/openshift/installer/pull/1850#issuecomment-503784200

/close

**Pull request closed without merging**